File: gem/lib/prosemaker.rb

require 'prosemaker/docpart'
require 'pp'

class ProseMaker
  # --------------------------------------------------
  # --------------------------------------------------
  attr_reader :data
  attr_reader :source
  # --------------------------------------------------
  def initialize
    @data = {}
    @source = ''
  end
  # --------------------------------------------------
  def doc
    @parts.map do |part|
      resolve_part part
    end.join
  end
  # --------------------------------------------------
  def data= new_data
    if new_data.respond_to? :[]
      @data = new_data
    else
      raise ArgumentError, 'new data value must be hash-like'
    end
  end
  # --------------------------------------------------
  def source= new_source
    if new_source.respond_to? :to_str
      @source = new_source.to_str
      @parts = []

      raw_parts = @source.split(/\[\[/)
      @parts = raw_parts.map do |raw_part|
        DocPart.new raw_part
      end
    else
      raise ArgumentError, 'new data value must be string-like'
    end
  end
  # --------------------------------------------------
  def resolve_content content
    content = content.to_s
    content = content + ' ' # add a trailing space so the split doesn't fail if the last thing is a var.

    len_sorted_varnames = data.keys.sort { |a, b| b.length <=> a.length }
    replacements = 1

    while replacements > 0 do
      replacements = 0
      len_sorted_varnames.each do |varname|
        bits = content.split('$$' + varname.to_s)
        replacements += bits.length - 1
        content = bits.join(data[varname].to_s)
      end
    end

    content.chop # remove the space we added at the start of this method
  end
  # --------------------------------------------------
  def resolve_condition condition
    condition != 'never'
  end
  # --------------------------------------------------
  def resolve_part part
    if resolve_condition part.condition
      resolve_content part.content
    else
      ''
    end
  end
  # --------------------------------------------------
  # --------------------------------------------------
end

File: egg/dwb.prosemaker/src/dwb/prosemaker/prosemaker.py

from docpart import DocPart

class ProseMaker(object):

    def __init__(self):
        self._data   = {}
        self._source = ''

    ## data ---------------------------------------------------------
    @property
    def data(self):
        """ The 'data' property """
        return self._data

    @data.setter
    def data(self, value):
        self._data = value
        return self._data

    @data.deleter
    def data(self):
        del self._data

    ## source -------------------------------------------------------
    @property
    def source(self):
        """ The 'source' property """
        return self._source

    @source.setter
    def source(self, value):
        self._source = value

        raw_parts = self._source.split(r"[[") ###]]" ### stupid Sublime syntax highlighter
        self._parts = [DocPart(raw_part) for raw_part in raw_parts]

        return self._source

    @source.deleter
    def source(self):
        del self._source

    ## doc ----------------------------------------------------------
    @property
    def doc(self):
        """ The 'doc' property """

        resolved_parts = [self.resolve_document_part(part) for part in self._parts]

        return(''.join(resolved_parts))

    # ---------------------------------------------------------------
    def resolve_document_part(self, doc_part):
        if self.resolve_condition(doc_part.condition):
            return self.resolve_content(doc_part.content)
        else:
            return ''

    # ---------------------------------------------------------------
    def resolve_condition(self, condition):
        condition = str(condition)
        if condition == 'always':
            return True
        if condition == 'never':
            return False

        # here's where the fancy parsing happens
        #
        #

        raise Exception('condition not understood')

    # ---------------------------------------------------------------
    def resolve_content(self, content):

        content = str(content) # force it into a string form (probably it's already a string)

        if len(self.data) > 0:

            content = content + ' ' # add a trailing space so the split doesn't fail if the last thing is a var.

            sorted_varnames = self.data.keys()
            sorted_varnames.sort(key=len, reverse=True) # sort longest first

            replacements = 1

            while (replacements > 0):
                replacements = 0
                for varname in sorted_varnames:
                    if ('$$' + varname) in content:
                        content = content.replace('$$' + varname, str(self.data[varname]))
                        replacements += 1

            content = content[:-1] # remove the space we added at the start of this method

        return content














